### DISCUSSION TITLE
Cox processes on GPU with very large datasets

### DISCUSSION THREAD

**Discussion by florinwalter** [cornellius-gp/gpytorch#1644] [Q&A] [answered]
Hi all!

I'm using the Pyro/GPyTorch low-level interface to implement a spatial Cox process, like [here](https://docs.gpytorch.ai/en/stable/examples/07_Pyro_Integration/Cox_Process_Example.html) but in 2D.
I train the model on a GPU, but this considerably limits the size of the dataset, since the whole dataset needs to be loaded on GPU at once.

Is there a workaround for this problem?

**Comment by MJimitater** [comment]
Hi @flcello and GPytorch Team! BTW thanks for this great library, I love it, keep up the good work!

This is no answer to your question @flcello , but rather the same problem.
Im trying to train GP classification with non-Gaussian likelihood. 
My training data set has approx. shape `(75000, 240)`. Moving this to a GPU with 6GB memory will throw an error that the memory ran out. So loading the whole dataset onto the GPU is not advisable, it needs to be "batchified" in some way, that only batches of data are moved to GPU and are thrown out once the training on them is done.
Is there some advise here please, some functionality of GPytorch or Pytorch to deal with this?
Thanks

**Comment by gpleiss** [comment, marked as answer]
@flcello and @MJimitater Since you are using variational inference, you can use a DataLoader (like in the [SVGP example](https://docs.gpytorch.ai/en/stable/examples/04_Variational_and_Approximate_GPs/SVGP_Regression_CUDA.html#Creating-a-DataLoader)) to only load part of your data into memory at any given time.

**Reply by MJimitater** [reply]
Awesome, thanks @gpleiss , it worked for me!

**Reply by florinwalter** [reply]
yes that works indeed, thanks! I was not sure whether minibatch training is applicable to a model with that particular likelihood, but tried it out and it nicely recovers the Poisson process intensity of my simulated data.

**Reply by michaelcao28** [reply]
hi @MJimitater, @flcello and @gpleiss, how did you (would you) write up your multi-output GP with poisson likelihood? I followed the examples [here](https://docs.gpytorch.ai/en/stable/examples/04_Variational_and_Approximate_GPs/SVGP_Multitask_GP_Regression.html) and [here](https://docs.gpytorch.ai/en/stable/examples/07_Pyro_Integration/Cox_Process_Example.html), but I can't seem to find a way to instantiate a likelihood outside of the class since the likelihood has to be multivariate. I could assume independent latent GPs for the intensity poisson parameter, with the mixing coefficients of these latent GPs being a learnable parameter, but I couldn't find examples of this. Any ideas or suggestions would be helpful :)
